concurrency/MpmcQueue.h

```diff
@@ -249,13 +282,21 @@ class MpmcQueue
 
     // Layout below must be maintained for performance. The non atomic member variables may not be place before read
     // cursor.
-    std::atomic<VersionedIndex> _readCursor;
-    uint32_t _cacheLineSeparator1[14];
-    std::atomic<VersionedIndex> _writeCursor;
-    uint32_t _cacheLineSeparator2[14];
-    const uint32_t _entriesPerCacheline;
-    const uint32_t _capacity;
-    const size_t _blockSize;
+    alignas(64) std::atomic<VersionedIndex> _readCursor;
+
+    // Special Cell state for when queue has zero size
+    // Null Entry is not written and it is only read it when queue is zero size. So it can live in same cache line of
+    // _readCursor without performance impact
+    const EntrySate _nullEntry = CellState::State::nullSlot;
+
+    alignas(64) std::atomic<VersionedIndex> _writeCursor;
+
+    // Although they are const. Cache can be invalidated for the reader when _writeCursor
+    // Perhaps is not a big issue (and even compile can optimize it this const values)
+    // but forcing a new cache line also ensures tha nothing is placed in this cache line after
+    // MpmcQueue boundary (which we can predict if can affect performance
+    // of _writeCursor)
+    alignas(64) const uint32_t _capacity;
```

Comment:
Having thought this through, this one should not be aligned. It is not read/written atomically and the value will be from whatever is in cache and it will not stall cpu pipeline to read it (which is ok as it is const). 
It is important that writecursor is protected from atomic operations in the next cache line, but that should  be resolved by all other objects doing atomic operations align their first atomic member variable. Preferably having the atomic one first to not waste memory.

Reply:
I remove his alignas. Then `_nullEntry`, `_capacity` and `_elements` are in the same cache line of _writeCursor to now waste space.

the `_writecursor` is protected from other atomic variables outside of this object because the `MpmcQueue` as an alignment requirement of 64 and the size needs to be a multiple of 64, so in this case the size is 128. Although there is no standardization on binary level, I think compilers don't have another option than forcing to be a multiple of aligment due to arrays (Example, if we have a MpmcQueue[N], accessing the memory position should `index * sizeof(MpmcQueue)` and that needs always be a 64 bytes aligned, then is ensured no other variables are added to the cache line outside of this object)